### ISSUE TITLE
Course Offering Schema

### ISSUE THREAD

**Issue by nananananate** [icssc-projects/peterportal-public-api#25]
The following is a brainstorming + Proposal of a course offering schema and beyond. Future discussion, ideas, and passionate debate can be put here. 

## What is a Course Offering?  
It is an "instance" of a course. Like a lab, lecture, discussion, etc. From what I've seen, some colleges/APIs call them "classes" or "sections". We'll be using **Course Offering** as the name in this discussion, although this may differ from our future implementation. 

For clarity, a college-agnostic definition of Course Offering fits the following criteria:  
* A consistent meeting time each week, lasting the duration of the term. Due to remote learning, this may show up as TBD. 
* A finite number of seats.  
* Define one or more instructors that are consistent for the course.    
* A Unique identifier (course offering code) defined by the college for every registration season.  
* A type (usually seminar, discussion, lecture, or lab).   
* A status for registration (usually FULL, OPEN, CLOSED, WAITLISTED).  

## The Problem  
For many course planning applications, it is important to have access to all course offerings of a course. More likely, an application would want to see all open lectures for a course.

Use cases:  
* Seeing which course offerings have the smallest waitlist
* Compare professor ratings for all course offerings in a course
* Tracking "hot" course offerings with the highest increase in enrollment in a given time period

**Comment by nananananate** [comment]
# Inspiration

This is a cool, albeit not perfect, implementation from [Northeastern's student-built deprecated API](https://github.com/sandboxnu/searchneu/blob/master/docs/API.md):  

```
{
    "score": 4.877671508066469,
    "type": "class",
    "class":
    {
        "lastUpdateTime": 1510519074966,
        "name": "Lab for CS 2510",
        "url": "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=201810&subj_in=CS&crse_in=2511&schd_in=%",
        "crns": ["11424", "12579", "13916", "14347"],
        "honors": false,
        "coreqs":
        {
            "type": "or",
            "values": [
            {
                "subject": "CS",
                "classId": "2510"
            }]
        },
        "maxCredits": 1,
        "minCredits": 1,
        "desc": "Accompanies CS 2510. Covers topics from the course through various experiments. 1.000 Lab hours",
        "classId": "2511",
        "prettyUrl": "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_course_detail?cat_term_in=201810&subj_code_in=CS&crse_numb_in=2511",
        "termId": "201810",
        "host": "neu.edu",
        "subject": "CS"
    },
    "sections": [
    {
        "seatsCapacity": 45,
        "seatsRemaining": 0,
        "waitCapacity": 0,
        "waitRemaining": 0,
        "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=11424",
        "crn": "11424",
        "meetings": [
        {
            "startDate": 17415,
            "endDate": 17506,
            "profs": ["Leena Razzaq"],
            "where": "West Village H 212",
            "type": "Class",
            "times":
            {
                "5": [
                {
                    "start": 42300,
                    "end": 48300
                }]
            },
            "allProfs": ["Leena Razzaq"]
        }],
        "lastUpdateTime": 1510519290251,
        "termId": "201810",
        "host": "neu.edu",
        "subject": "CS",
        "classId": "2511",
        "campus": "Boston"
    },
    {
        "seatsCapacity": 45,
        "seatsRemaining": 1,
        "waitCapacity": 0,
        "waitRemaining": 0,
        "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=12579",
        "crn": "12579",
        "meetings": [
        {
            "startDate": 17415,
            "endDate": 17506,
            "profs": ["Leena Razzaq"],
            "where": "West Village H 212",
            "type": "Class",
            "times":
            {
                "5": [
                {
                    "start": 35400,
                    "end": 41400
                }]
            },
            "allProfs": ["Leena Razzaq"]
        }],
        "lastUpdateTime": 1510519290243,
        "termId": "201810",
        "host": "neu.edu",
        "subject": "CS",
        "classId": "2511",
        "campus": "Boston"
    },
    {
        "seatsCapacity": 50,
        "seatsRemaining": 21,
        "waitCapacity": 0,
        "waitRemaining": 0,
        "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=13916",
        "crn": "13916",
        "meetings": [
        {
            "startDate": 17415,
            "endDate": 17506,
            "profs": ["Leena Razzaq"],
            "where": "West Village H 210",
            "type": "Class",
            "times":
            {
                "5": [
                {
                    "start": 48900,
                    "end": 54900
                }]
            },
            "allProfs": ["Leena Razzaq"]
        }],
        "lastUpdateTime": 1510519290247,
        "termId": "201810",
        "host": "neu.edu",
        "subject": "CS",
        "classId": "2511",
        "campus": "Boston"
    },
    {
        "seatsCapacity": 50,
        "seatsRemaining": 30,
        "waitCapacity": 0,
        "waitRemaining": 0,
        "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=14347",
        "crn": "14347",
        "meetings": [
        {
            "startDate": 17415,
            "endDate": 17506,
            "profs": ["Leena Razzaq"],
            "where": "West Village H 210",
            "type": "Class",
            "times":
            {
                "5": [
                {
                    "start": 55500,
                    "end": 61500
                }]
            },
            "allProfs": ["Leena Razzaq"]
        }],
        "lastUpdateTime": 1510519290249,
        "termId": "201810",
        "host": "neu.edu",
        "subject": "CS",
        "classId": "2511",
        "campus": "Boston"
    }]
}
```

**Comment by ChaseC99** [comment]
Initial Course Offering Type implemented in #43 
```graphql
type CourseOffering {
  year: String,
  quarter: String,
  code: float,
  section: String,
  type: String,
  instructor: String,  // map to professorType
  course: courseType 
}
```
